File: legacy/listening-word/scripts/download_audio.rb
#!/usr/bin/env ruby

require "json"
require "fileutils"
require "net/http"
require "uri"
require "optparse"
require "thread"
require "time"

SCRIPT_DIR = __dir__
PROJECT_ROOT = File.expand_path("..", SCRIPT_DIR)
DEFAULT_SOURCE = File.expand_path("王璐语料库_源码.html", PROJECT_ROOT)
DEFAULT_OUTPUT = File.expand_path("assets/audio", PROJECT_ROOT)
DEFAULT_BASE_URL = "http://www.1kao.com.cn/iSpell/Spell/audio"

def normalize_word_file_name(word)
  word.to_s.strip.gsub(/[\\\/:*?"<>|]/, "_").gsub(/\s+/, "_")
end

def parse_options
  options = {
    source: DEFAULT_SOURCE,
    output: DEFAULT_OUTPUT,
    base_url: DEFAULT_BASE_URL,
    chapter: nil,
    concurrency: 6,
    retry: 3,
    timeout: 15,
    limit: 0,
    force: false,
    dry_run: false
  }

  parser = OptionParser.new do |opts|
    opts.banner = "用法: ruby download_audio.rb [options]"

    opts.on("--source PATH", "语料 HTML 源文件") { |value| options[:source] = File.expand_path(value) }
    opts.on("--output PATH", "音频输出目录") { |value| options[:output] = File.expand_path(value) }
    opts.on("--base-url URL", "音频基地址") { |value| options[:base_url] = value.sub(%r{/$}, "") }
    opts.on("--chapter ID", "仅下载指定章节") { |value| options[:chapter] = value.strip }
    opts.on("--concurrency N", Integer, "并发下载数，默认 6") { |value| options[:concurrency] = [value, 1].max }
    opts.on("--retry N", Integer, "失败重试次数，默认 3") { |value| options[:retry] = [value, 1].max }
    opts.on("--timeout N", Integer, "单文件超时秒数，默认 15") { |value| options[:timeout] = [value, 1].max }
    opts.on("--limit N", Integer, "仅处理前 N 个文件") { |value| options[:limit] = [value, 0].max }
    opts.on("--force", "覆盖已存在文件") { options[:force] = true }
    opts.on("--dry-run", "只输出计划，不实际下载") { options[:dry_run] = true }
    opts.on("-h", "--help", "显示帮助") do
      puts opts
      exit 0
    end
  end

  parser.parse!
  options
end

def extract_chapter_word_sets(source_html)
  match = source_html.match(/const CHAPTER_WORD_SETS = (\{[\s\S]*?\n\});\n\s*const EXPORT_FILE_PREFIX/)
  raise "未能从 HTML 中提取 CHAPTER_WORD_SETS 对象。" unless match

  JSON.parse(match[1])
end

def build_tasks(chapter_word_sets, options)
  entries = chapter_word_sets.sort_by { |chapter_id, _chapter| chapter_id.to_i }
  entries = entries.select { |chapter_id, _chapter| chapter_id == options[:chapter] } if options[:chapter]

  if options[:chapter] && entries.empty?
    raise "章节 #{options[:chapter]} 不存在。"
  end

  tasks = entries.flat_map do |chapter_id, chapter|
    chapter.fetch("words").map do |word|
      encoded_word = URI.encode_www_form_component(word)
      word_file_name = "#{normalize_word_file_name(word)}.mp3"
      {
        chapter_id: chapter_id,
        chapter_title: chapter.fetch("title"),
        word: word,
        url: "#{options[:base_url]}/#{URI.encode_www_form_component(chapter_id)}/#{encoded_word}.mp3",
        original_relative_path: File.join(chapter_id, "#{encoded_word}.mp3"),
        original_output_path: File.join(options[:output], chapter_id, "#{encoded_word}.mp3"),
        word_relative_path: File.join("by-word", chapter_id, word_file_name),
        word_output_path: File.join(options[:output], "by-word", chapter_id, word_file_name)
      }
    end
  end

  options[:limit] > 0 ? tasks.first(options[:limit]) : tasks
end

def write_manifest(tasks, output_dir)
  FileUtils.mkdir_p(output_dir)
  manifest = {
    generatedAt: Time.now.iso8601,
    total: tasks.length,
    files: tasks.map do |task|
      {
        chapterId: task[:chapter_id],
        chapterTitle: task[:chapter_title],
        word: task[:word],
        url: task[:url],
        file: task[:original_relative_path],
        originalFile: task[:original_relative_path],
        wordFile: task[:word_relative_path]
      }
    end
  }
  File.write(File.join(output_dir, "manifest.json"), JSON.pretty_generate(manifest))
end

def file_ready?(file_path)
  File.file?(file_path) && File.size?(file_path)
end

def fetch_to_file(url_string, target_path, timeout_seconds, limit = 5)
  raise "重定向次数过多" if limit <= 0

  uri = URI.parse(url_string)
  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = uri.scheme == "https"
  http.open_timeout = timeout_seconds
  http.read_timeout = timeout_seconds

  request = Net::HTTP::Get.new(uri.request_uri)

  http.request(request) do |response|
    case response
    when Net::HTTPSuccess
      FileUtils.mkdir_p(File.dirname(target_path))
      temp_path = "#{target_path}.part"
      File.open(temp_path, "wb") do |file|
        response.read_body { |chunk| file.write(chunk) }
      end
      File.rename(temp_path, target_path)
    when Net::HTTPRedirection
      location = response["location"]
      raise "缺少重定向地址" if location.to_s.empty?

      redirected_url = URI.join(url_string, location).to_s
      fetch_to_file(redirected_url, target_path, timeout_seconds, limit - 1)
    else
      raise "HTTP #{response.code}"
    end
  end
rescue StandardError
  FileUtils.rm_f("#{target_path}.part")
  raise
end

def download_with_retry(task, options)
  attempts = options[:retry]

  1.upto(attempts) do |attempt|
    begin
      fetch_to_file(task[:url], task[:original_output_path], options[:timeout])
      return [true, attempt, nil]
    rescue StandardError => error
      return [false, attempt, error] if attempt == attempts
    end
  end

  [false, attempts, RuntimeError.new("UNKNOWN")]
end

def copy_word_named_file(task)
  FileUtils.mkdir_p(File.dirname(task[:word_output_path]))
  FileUtils.cp(task[:original_output_path], task[:word_output_path])
end

options = parse_options
source_html = File.read(options[:source])
chapter_word_sets = extract_chapter_word_sets(source_html)
tasks = build_tasks(chapter_word_sets, options)

write_manifest(tasks, options[:output])

puts "源文件: #{options[:source]}"
puts "输出目录: #{options[:output]}"
puts "计划处理: #{tasks.length} 个音频文件"
puts "并发数: #{options[:concurrency]}，重试次数: #{options[:retry]}"

if tasks.empty?
  puts "没有可处理的音频任务。"
  exit 0
end

if options[:dry_run]
  puts "dry-run 模式，不执行实际下载。"
  exit 0
end

downloaded = 0
skipped = 0
synced = 0
failed = 0
mutex = Mutex.new
queue = Queue.new
tasks.each_with_index { |task, index| queue << [task, index] }
started_at = Time.now

workers = Array.new([options[:concurrency], tasks.length].min) do
  Thread.new do
    loop do
      task, index = queue.pop(true)
      prefix = "[#{index + 1}/#{tasks.length}] #{task[:chapter_id]} #{task[:word]}"

      has_original = file_ready?(task[:original_output_path])
      has_word_named = file_ready?(task[:word_output_path])

      if !options[:force] && has_original
        if has_word_named
          mutex.synchronize do
            skipped += 1
            puts "#{prefix} -> 两份文件已存在，跳过"
          end
          next
        end

        begin
          copy_word_named_file(task)
          mutex.synchronize do
            synced += 1
            puts "#{prefix} -> 已补齐单词命名文件"
          end
          next
        rescue StandardError => error
          mutex.synchronize do
            failed += 1
            warn "#{prefix} -> 补齐单词命名文件失败: #{error.message}"
          end
          next
        end
      end

      ok, attempts, error = download_with_retry(task, options)
      mutex.synchronize do
        if ok
          begin
            copy_word_named_file(task)
            downloaded += 1
            puts "#{prefix} -> 下载完成并生成两份文件 (尝试 #{attempts} 次)"
          rescue StandardError => copy_error
            failed += 1
            warn "#{prefix} -> 下载成功但生成单词命名文件失败: #{copy_error.message}"
          end
        else
          failed += 1
          warn "#{prefix} -> 下载失败: #{error.message}"
        end
      end
    rescue ThreadError
      break
    end
  end
end

workers.each(&:join)

duration = (Time.now - started_at).round(1)
puts "完成: 下载 #{downloaded}，补齐 #{synced}，跳过 #{skipped}，失败 #{failed}，耗时 #{duration}s"
exit(1) if failed.positive?

File: legacy/listening-word/scripts/download_audio.js
#!/usr/bin/env node

const fs = require("fs");
const path = require("path");
const vm = require("vm");
const http = require("http");
const https = require("https");

const SCRIPT_DIR = __dirname;
const PROJECT_ROOT = path.resolve(SCRIPT_DIR, "..");
const DEFAULT_SOURCE = path.resolve(PROJECT_ROOT, "王璐语料库_源码.html");
const DEFAULT_OUTPUT = path.resolve(PROJECT_ROOT, "assets", "audio");
const DEFAULT_CONCURRENCY = 6;
const DEFAULT_RETRY = 3;
const DEFAULT_TIMEOUT_MS = 15000;
const DEFAULT_BASE_URL = "http://www.1kao.com.cn/iSpell/Spell/audio";

function normalizeWordFileName(word) {
  return String(word)
    .trim()
    .replace(/[\\/:*?"<>|]/g, "_")
    .replace(/\s+/g, "_");
}

function parseArgs(argv) {
  const options = {
    source: DEFAULT_SOURCE,
    output: DEFAULT_OUTPUT,
    concurrency: DEFAULT_CONCURRENCY,
    retry: DEFAULT_RETRY,
    timeoutMs: DEFAULT_TIMEOUT_MS,
    baseUrl: DEFAULT_BASE_URL,
    chapter: "",
    force: false,
    limit: 0,
    dryRun: false,
  };

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];

    if (arg === "--source") {
      options.source = path.resolve(argv[index + 1] || "");
      index += 1;
      continue;
    }

    if (arg === "--output") {
      options.output = path.resolve(argv[index + 1] || "");
      index += 1;
      continue;
    }

    if (arg === "--concurrency") {
      options.concurrency = Math.max(1, Number.parseInt(argv[index + 1] || String(DEFAULT_CONCURRENCY), 10) || DEFAULT_CONCURRENCY);
      index += 1;
      continue;
    }

    if (arg === "--retry") {
      options.retry = Math.max(1, Number.parseInt(argv[index + 1] || String(DEFAULT_RETRY), 10) || DEFAULT_RETRY);
      index += 1;
      continue;
    }

    if (arg === "--timeout") {
      options.timeoutMs = Math.max(1000, Number.parseInt(argv[index + 1] || String(DEFAULT_TIMEOUT_MS), 10) || DEFAULT_TIMEOUT_MS);
      index += 1;
      continue;
    }

    if (arg === "--base-url") {
      options.baseUrl = String(argv[index + 1] || DEFAULT_BASE_URL).replace(/\/$/, "");
      index += 1;
      continue;
    }

    if (arg === "--chapter") {
      options.chapter = String(argv[index + 1] || "").trim();
      index += 1;
      continue;
    }

    if (arg === "--limit") {
      options.limit = Math.max(0, Number.parseInt(argv[index + 1] || "0", 10) || 0);
      index += 1;
      continue;
    }

    if (arg === "--force") {
      options.force = true;
      continue;
    }

    if (arg === "--dry-run") {
      options.dryRun = true;
      continue;
    }

    if (arg === "--help" || arg === "-h") {
      printHelp();
      process.exit(0);
    }

    throw new Error(`未知参数：${arg}`);
  }

  return options;
}

function printHelp() {
  console.log(`用法: node download_audio.js [options]

参数:
  --source <path>        语料 HTML 源文件，默认 ../王璐语料库_源码.html
  --output <path>        音频输出目录，默认 ../assets/audio
  --chapter <id>         仅下载指定章节，例如 31
  --concurrency <n>      并发下载数，默认 6
  --retry <n>            单文件失败重试次数，默认 3
  --timeout <ms>         单文件超时毫秒数，默认 15000
  --base-url <url>       音频基地址，默认 http://www.1kao.com.cn/iSpell/Spell/audio
  --limit <n>            仅处理前 n 个文件，便于测试
  --force                覆盖已存在文件
  --dry-run              只输出下载计划，不实际下载
  --help, -h             显示帮助
`);
}

function extractChapterWordSets(sourceHtml) {
  const match = sourceHtml.match(/const CHAPTER_WORD_SETS = (\{[\s\S]*?\n\});\n\s*const EXPORT_FILE_PREFIX/);
  if (!match) {
    throw new Error("未能从 HTML 中提取 CHAPTER_WORD_SETS 对象。");
  }

  return vm.runInNewContext(`(${match[1]})`, Object.create(null));
}

function buildTasks(chapterWordSets, options) {
  const entries = Object.entries(chapterWordSets)
    .sort((left, right) => Number(left[0]) - Number(right[0]));

  const filteredEntries = options.chapter
    ? entries.filter(([chapterId]) => chapterId === options.chapter)
    : entries;

  if (options.chapter && !filteredEntries.length) {
    throw new Error(`章节 ${options.chapter} 不存在。`);
  }

  const tasks = [];
  for (const [chapterId, chapter] of filteredEntries) {
    for (const word of chapter.words) {
      const encodedWord = encodeURIComponent(word);
      const wordFileName = `${normalizeWordFileName(word)}.mp3`;
      tasks.push({
        chapterId,
        chapterTitle: chapter.title,
        word,
        url: `${options.baseUrl}/${encodeURIComponent(chapterId)}/${encodedWord}.mp3`,
        originalOutputPath: path.join(options.output, chapterId, `${encodedWord}.mp3`),
        originalRelativePath: path.join(chapterId, `${encodedWord}.mp3`),
        wordOutputPath: path.join(options.output, "by-word", chapterId, wordFileName),
        wordRelativePath: path.join("by-word", chapterId, wordFileName),
      });
    }
  }

  return options.limit > 0 ? tasks.slice(0, options.limit) : tasks;
}

function ensureDir(dirPath) {
  fs.mkdirSync(dirPath, { recursive: true });
}

function fileExists(filePath) {
  try {
    const stats = fs.statSync(filePath);
    return stats.isFile() && stats.size > 0;
  } catch (_error) {
    return false;
  }
}

function fetchToFile(urlString, targetPath, timeoutMs) {
  return new Promise((resolve, reject) => {
    const url = new URL(urlString);
    const client = url.protocol === "https:" ? https : http;
    const request = client.get(url, (response) => {
      if (response.statusCode && response.statusCode >= 300 && response.statusCode < 400 && response.headers.location) {
        response.resume();
        const redirectedUrl = new URL(response.headers.location, urlString).toString();
        fetchToFile(redirectedUrl, targetPath, timeoutMs).then(resolve).catch(reject);
        return;
      }

      if (response.statusCode !== 200) {
        response.resume();
        reject(new Error(`HTTP ${response.statusCode || "UNKNOWN"}`));
        return;
      }

      ensureDir(path.dirname(targetPath));
      const tempPath = `${targetPath}.part`;
      const fileStream = fs.createWriteStream(tempPath);

      response.pipe(fileStream);

      fileStream.on("finish", () => {
        fileStream.close((closeError) => {
          if (closeError) {
            reject(closeError);
            return;
          }

          fs.rename(tempPath, targetPath, (renameError) => {
            if (renameError) {
              reject(renameError);
              return;
            }
            resolve();
          });
        });
      });

      fileStream.on("error", (streamError) => {
        response.destroy();
        fs.rm(tempPath, { force: true }, () => reject(streamError));
      });
    });

    request.setTimeout(timeoutMs, () => {
      request.destroy(new Error(`TIMEOUT ${timeoutMs}ms`));
    });

    request.on("error", (error) => {
      fs.rm(`${targetPath}.part`, { force: true }, () => reject(error));
    });
  });
}

async function downloadWithRetry(task, options) {
  const maxAttempts = options.retry;

  for (let attempt = 1; attempt <= maxAttempts; attempt += 1) {
    try {
      await fetchToFile(task.url, task.originalOutputPath, options.timeoutMs);
      return { ok: true, attempts: attempt };
    } catch (error) {
      if (attempt === maxAttempts) {
        return { ok: false, attempts: attempt, error };
      }
    }
  }

  return { ok: false, attempts: maxAttempts, error: new Error("UNKNOWN") };
}

async function runPool(items, concurrency, worker) {
  let cursor = 0;
  const results = new Array(items.length);

  async function consume() {
    while (cursor < items.length) {
      const currentIndex = cursor;
      cursor += 1;
      results[currentIndex] = await worker(items[currentIndex], currentIndex);
    }
  }

  const workers = Array.from({ length: Math.min(concurrency, items.length || 1) }, () => consume());
  await Promise.all(workers);
  return results;
}

function writeManifest(tasks, outputDir) {
  const manifest = {
    generatedAt: new Date().toISOString(),
    total: tasks.length,
    files: tasks.map((task) => ({
      chapterId: task.chapterId,
      chapterTitle: task.chapterTitle,
      word: task.word,
      url: task.url,
      file: task.originalRelativePath,
      originalFile: task.originalRelativePath,
      wordFile: task.wordRelativePath,
    })),
  };

  ensureDir(outputDir);
  fs.writeFileSync(path.join(outputDir, "manifest.json"), JSON.stringify(manifest, null, 2));
}

function copyWordNamedFile(task) {
  ensureDir(path.dirname(task.wordOutputPath));
  fs.copyFileSync(task.originalOutputPath, task.wordOutputPath);
}

async function main() {
  const options = parseArgs(process.argv.slice(2));
  const sourceHtml = fs.readFileSync(options.source, "utf8");
  const chapterWordSets = extractChapterWordSets(sourceHtml);
  const tasks = buildTasks(chapterWordSets, options);

  ensureDir(options.output);
  writeManifest(tasks, options.output);

  console.log(`源文件: ${options.source}`);
  console.log(`输出目录: ${options.output}`);
  console.log(`计划处理: ${tasks.length} 个音频文件`);
  console.log(`并发数: ${options.concurrency}，重试次数: ${options.retry}`);

  if (!tasks.length) {
    console.log("没有可处理的音频任务。");
    return;
  }

  if (options.dryRun) {
    console.log("dry-run 模式，不执行实际下载。");
    return;
  }

  let downloaded = 0;
  let skipped = 0;
  let synced = 0;
  let failed = 0;

  const startedAt = Date.now();

  await runPool(tasks, options.concurrency, async (task, index) => {
    const prefix = `[${index + 1}/${tasks.length}] ${task.chapterId} ${task.word}`;

    const hasOriginal = fileExists(task.originalOutputPath);
    const hasWordNamed = fileExists(task.wordOutputPath);

    if (!options.force && hasOriginal) {
      if (hasWordNamed) {
        skipped += 1;
        console.log(`${prefix} -> 两份文件已存在，跳过`);
        return;
      }

      try {
        copyWordNamedFile(task);
        synced += 1;
        console.log(`${prefix} -> 已补齐单词命名文件`);
        return;
      } catch (error) {
        failed += 1;
        console.error(`${prefix} -> 补齐单词命名文件失败: ${error.message || "UNKNOWN"}`);
        return;
      }
    }

    const result = await downloadWithRetry(task, options);
    if (result.ok) {
      try {
        copyWordNamedFile(task);
        downloaded += 1;
        console.log(`${prefix} -> 下载完成并生成两份文件 (尝试 ${result.attempts} 次)`);
        return;
      } catch (error) {
        failed += 1;
        console.error(`${prefix} -> 下载成功但生成单词命名文件失败: ${error.message || "UNKNOWN"}`);
        return;
      }
    }

    failed += 1;
    console.error(`${prefix} -> 下载失败: ${result.error ? result.error.message : "UNKNOWN"}`);
  });

  const durationSeconds = ((Date.now() - startedAt) / 1000).toFixed(1);
  console.log(`完成: 下载 ${downloaded}，补齐 ${synced}，跳过 ${skipped}，失败 ${failed}，耗时 ${durationSeconds}s`);

  if (failed > 0) {
    process.exitCode = 1;
  }
}

main().catch((error) => {
  console.error(error.message || error);
  process.exit(1);
});